# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmBaselineCompile.cpp
# Commit: af2e64962ccf
# Full Hash: af2e64962ccfbdef97762dccbba9dda2ddb03b87
# Author: Andy Wingo <wingo@igalia.com>
# Date: 2020-04-10 02:59:09
# Description:
#   Bug 1628417 - Fix multi-value br stack value shuffling in baseline compiler r=bbouvier
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D70327
# ==============================================================================

diff -r 3ecec71f8dcf -r af2e64962ccf js/src/wasm/WasmBaselineCompile.cpp
--- a/js/src/wasm/WasmBaselineCompile.cpp	Thu Apr 09 22:35:35 2020 +0300
+++ b/js/src/wasm/WasmBaselineCompile.cpp	Thu Apr 09 13:43:59 2020 +0000
@@ -4361,18 +4361,20 @@
     if (ABIResultIter::HasStackResults(type)) {
       MOZ_ASSERT(stk_.length() >= type.length());
       ABIResultIter iter(type);
-      for (ABIResultIter iter(type); !iter.done(); iter.next()) {
+      for (; !iter.done(); iter.next()) {
 #ifdef DEBUG
         const ABIResult& result = iter.cur();
         const Stk& v = stk_[stk_.length() - iter.index() - 1];
         MOZ_ASSERT(v.isMem() == result.onStack());
 #endif
       }
+
       stackResultBytes = iter.stackBytesConsumedSoFar();
-
-      if (stackResultBytes) {
-        // Find a free GPR to use when shuffling stack values.  If none is
-        // available, push ReturnReg and restore it after we're done.
+      MOZ_ASSERT(stackResultBytes > 0);
+
+      if (srcHeight != destHeight) {
+        // Find a free GPR to use when shuffling stack values.  If none
+        // is available, push ReturnReg and restore it after we're done.
         bool saved = false;
         RegPtr temp = ra.needTempPtr(RegPtr(ReturnReg), &saved);
         fr.shuffleStackResultsTowardFP(srcHeight, destHeight, stackResultBytes,
